### ISSUE TITLE
Proposal for weaving contracts onto functions

### ISSUE THREAD

**Issue by blerner** [brownplt/pyret-lang#1114]
We've left this as a TODO item for ages, and it prompted a discussion about how contract lines are/aren't comments...  We've left this open because it's not immediately clear what it means to weave annotations in, so here's a concrete proposal.

1. *Where are annotations allowed?*  The challenge here is matching the identifiers shown in the contracts to the resolved `Name`s in definitions, especially since they appear in the wrong order, appear before name resolution occurs, and we have to handle something to do with shadowing.  I suggest we allow two positions: for any given file, contracts should appear _either_ consecutively in a bunch of contracts at the top of the file (just below any provide/import declarations), _or_ we allow contracts to appear immediately before the definition group in which their name appears.  (I have to quibble about "definition group" in order to accommodate let-rec'ed functions.)  This solves the name-resolution problem: if presented just before a definition group, then it's' "whatever `Name` comes next".  If presented at the top of the file, then it's "whatever the final binding of that `Name` is", exactly how `provide` works.

2. *What to do about imports?*  I think we should _wrap_ imported functions with a new lambda that does two things: it enforces the annotation as specified here, and intercepts any general-type-or-arity-mismatch errors from the underlying and complains about an annotation-mismatch (a new kind of error) between the purported contract and the underlying signature of the function.  In other words, if a student writes
```
circle :: Number -> Image
```
which is blatantly wrong, then this contract should desugar to (a pseudocode mishmash of Pyret and JS)
```
shadow circle = lam(n :: Number) -> Image:
  try {
    ans = circle(n)
    if (ans is-a Image) { return ans }
    else { raise contract-return-mismatch(ans, Image, srcloc of contract) }
  } catch(e) {
    raise contract-mismatch(e, srcloc of contract);
  }
end
```
The idea being, the underlying imported function is "trusted" as ground-truth, and the annotation here is more likely to be wrong.  The `contract-mismatch(e, srcloc)` idea is to be rendered as "The actual function expected an arity ### or arguments of type ###, but the contract here stipulated something else: double-check your contract."  The `contract-return-mismatch(ans, Image, srcloc)` idea is to handle the anonymous return value more flexibly.

3. *What to do about redundant contracts?* If a student writes
```
square-num :: Number -> Number
fun square-num(n :: Number) -> Number:
  n * n
end
```
should that be a well-formedness error because the parameter is explicitly annotated anyway?  What if the contract and the annotation disagree?  I'm inclined to say that specifying both a contract and an annotated function definition should be a well-fordmedness error, *or* to make it be a wrapped-lambda like the imported case above, with a preference for the well-formedness error.

Thoughts?

**Comment by shriram** [comment]
I like all of this except for feeling uncomfortable about the
well-formedness error for #3. It feels like the easy way out of a harder
problem. Why not do the wrapping? That way there's no harm to writing it
twice, so long as they match. It would also be easier for a code-generator
to know it doesn't have to keep track of where it's put annotations (though
in general generated code may want to skip well-formedness-checking…).

The only downside is that there will be a performance penalty, and it'll be
subtle to figure out that it's happening.

**Comment by blerner** [comment]
@shriram from last week: "Performance!"
@shriram from today: "Why not do the wrapping?"

;-)

I think the well-formedness error here isn't quite so bad, since the contract and the annotation are both manifestly present, and so saying "one of these is redundant" is kinda "duh".  We could also compromise and say "if they're syntactically identical annotations, then no well-formedness error; if they're not, then syntactic well-formedness error"

**Comment by shriram** [comment]
I'm okay with well-formedness. I brought up the performance point myself!

**Comment by blerner** [comment]
Joe raises a good point: the `try/catch` I mention in option 2 is not easily feasible/implementable until we have a way to distinguish "errors raised by the callee" from "errors transitively raised by the callee".  So this requires some other infrastructure work, to distinguish locally-raised from transitively-raised errors...

**Comment by blerner** [comment]
Ok -- @jpolitz , @shriram , are you ok if I start implements parts 1 and 3 (using well-formedness)?  I think this is something we *need* to get done soon if we want teachers to use it, since it's already causing confusion in the Physics workshop...  Part 2 has to wait for other work (Joe had a suggestion on Slack about this, I think, which has languished for a while as a low-priority item).  One hacktastic approach to it is to measure the number of lines of the stack trace of the caught exception, vs the height of one created in the catch block, and iff they differ by one, then it's an exception thrown directly by the callee.

**Comment by jpolitz** [comment]
I think, for a first impl, make imports produce a well-formedness/shadowing
kind of error. That allows us to make progress on this, which I think is
important, without picking a semantics we are tied to forever.

**Comment by blerner** [comment]
Well the other alternative, even lighter-weight and potentially easier to use with the DR, is to ignore contracts on imported names, so students can still write the contracts without any error feedback.

**Comment by jpolitz** [comment]
I'd rather not introduce a kind of program that will work, and later will
behave oddly or produce an error, than the alternative.

So I'd prefer to add the error (even just re-use the shadowing
infrastructure) than let it silently "succeed"

**Comment by blerner** [comment]
I don't see it that way: We're not introducing anything.  *Currently* they work, because contracts are silently ignored everywhere (assuming they parse and have no unbound type identifiers).  I'm saying leave contracts for imports alone and continue to ignore them.  I think this is probably better than the alternative of explaining "Write down the contract for these functions your worksheet.  You can copy your worksheet into your code...but watch out, don't copy contracts for any of the functions we give you!  Why?  Well...because."

**Comment by jpolitz** [comment]
IMO, we are introducing something. No curriculum has made use of standalone
contract syntax, which sort of awkwardly exist as a weird comment syntax.

It was a _mistake_ that I typed contracts for existing functions into the
editor.

I wish Shriram _hadn't_ pushed for using this syntax before it had a
meaning.

We didn't have a semantics in mind, and we haphazardly introduced syntax,
and now we're using that haphazard introduction, including my mistake, to
drive semantic design.

**Comment by shriram** [comment]
> No curriculum has made use of standalone contract syntax

I have, extensively. And it's really important to make it work from a DR perspective:

- write your contract
- write your examples

which is very, very awkward when your "contract" is of the form
```
fun f(x :: Number, y :: Number) -> String: …
```
What is that? It's not syntactically valid, it's not even clear what form I should use (did I need the `…`? should there have been an `end`?) — and students really want to know that they're writing in the right form.

**Comment by jpolitz** [comment]
> I have, extensively. And it's really important to make it work from a DR perspective

Of course it's important to write them down from a DR perspective. I'm not arguing that standalone contracts aren't important, or that we shouldn't do this.

I'm frustrated that we pushed syntax that we haven't given a meaning to yet, and now we're trying to back-fit semantics onto (effectively) comments. We should be in the contented position of picking what we want them to mean, not saying “well, at this one workshop/in this one set of notes we wrote it this way, and we don't want to cause errors with it now that the code exists, therefore the existing syntax needs to work in exactly that context.”

**Comment by shriram** [comment]
Sure, I'm fine with not overly adhering to what happened at BS:P last week.
I viewed Ben using that as a guideline, not as a cudgel.

**Comment by jpolitz** [comment]
Right, but my argument is that we shouldn't be in the habit of writing uninterpreted stuff into files (or providing/encouraging uninterpreted syntax). So we shouldn't go with the route that has this syntax mean something for some functions, and nothing for some functions. If it has no meaning, it should be an error.

**Comment by jpolitz** [comment]
To be clear – with the change that it's a shadowing error to provide a contract for an imported name, I really like this proposal.

**Comment by shriram** [comment]
>
> To be clear – with the change that it's a shadowing error to provide a
> contract for an imported name, I really like this proposal.
>

I understand why this is needed for now. It is disappointing that the
"contract sheet" at the back of the workbook can't be written into the
program file. Would you object to this being lifted someday?

**Comment by jpolitz** [comment]
That's a good point. I'm not opposed to giving contracts for imported names meaning.

I just think there are many interpretations of it, and picking/designing one is going to be a challenge (it will also affect the type-checker, it has to do with nested errors, or computation that involves comparing annotations at runtime, or a performance hit from extra wrapping, etc).

As a result, it's worth deferring it, and getting the syntax in for user-defined functions to support DR-style development, which would be a huge win.

**Comment by shriram** [comment]
I'm fine with *deferring* it for these reasons.

Ben, are you in agreement here?

**Comment by blerner** [comment]
That was exactly [my comment](https://github.com/brownplt/pyret-lang/issues/1114#issuecomment-321143867) from before:  "I'm saying leave contracts for imports alone and continue to ignore them."

So yes, I agree we should implement 1 and 3, and defer implementing 2 until the needed infrastructure is there, but *allow and silently ignore for now* contracts on imported names.

**Comment by shriram** [comment]
Can it at least not be _silent_? Can it say "ignoring shadowed contract at line …"?

**Comment by blerner** [comment]
We have no notion right now of _warnings_ in Pyret -- even well-formedness complaints are fatal, program-blocking errors.  Are you suggesting that we add this new form of feedback?  If so that opens up a whole can of design worms...

**Comment by shriram** [comment]
Argh. Okay, I'll take an error in return for getting this to work at least for new functions.

**Comment by blerner** [comment]
Ok, then this is gonna have to be a carefully worded message. "Pyret doesn't currently allow writing or enforcing contracts for imported functions. Please comment this contract out, for now." Or something... I'm not sure what wording you want, so I'll use the text above until you all hash out what you'd prefer it to say.

**Comment by shriram** [comment]
Just curious: why do you say "or enforcing"? If the function had a contract
at its definition, would it not be enforced?

**Comment by blerner** [comment]
Lemme add second-person language to that message (which we've tried to eliminate): "Pyret doesn't currently allow *you* to write or enforce contracts for imported functions that *someone else* wrote.  *You* are allowed to write and enforce contracts on functions that *you* write."

*You* tell *me* how to phrase this appropriately ;-)

**Comment by jpolitz** [comment]
How about: functions can only have one contract, and the contract for
circle is defined in the image library.

**Comment by shriram** [comment]
On Thu, Aug 10, 2017 at 11:09 AM, Joe Politz <notifications@github.com>
wrote:

> How about: functions can only have one contract, and the contract for
> circle is defined in the image library.


I like this! Small amendment:

Contracts for functions can be defined only once, and the contract for … is
already defined in ….

[The way Joe has written it, one could say, "it *does* have only one
contract; I just wrote the same contract a second time!")

**Comment by blerner** [comment]
Hmm, some wrinkles in these plans. First, my spec for 1 was ambiguous: contracts that appear before the first function definition could be considered "for the final values of the file" (i.e. whatever names a provide statement would export), or "for the upcoming definition group".  @shriram, @schanzer, @jpolitz, need your suggestions here of which interpretation I should use. I think we should prefer the one that works better with BS, but I'm not sure which one that is.


And the whole notion of "final values of the file" is problematic, because of phase-ordering problems.  (Implementation headaches ahead.) Basically: initially, we have a syntactic block structure for our programs. Then, desugar-scope converts that into lexically nested let expressions. Next, resolve-names determines what Name goes with which identifier.  I need those Names to connect whole-file contracts to their definitions, but I need the block structure to connect middle-of-the-file contracts to their subsequent definitions. So somehow I need to run this contract weaving process at two times in the compilation pipeline, and I don't quite see how to do it yet. @jpolitz, ideas? (Can discuss offline from this issue discussion.)

Lastly, an unfortunate syntactic hiccup:
```
foo :: (A -> B)
foo = lam(a): b end

bar :: (A -> B) = lam(a): b end
```
Under our current interpretation, `foo` will have its annotation woven in, but `bar` will merely be checked that it is a function. Is this worth fixing (assuming I get the rest of weaving working)?

**Comment by jpolitz** [comment]
I think BS doesn't typically care about exports, so preferring to tie a contract to a function definition is preferred. Folks working with exports can do a little extra work.

Implementation-wise, the current name resolution uses a bit of state to grab the "innermost" environment that's calculated in resolve-scope and make it available outside the traversal. Is that a useful idea to highlight – that we could collect all the name information, maybe with an extra flag that says it was a top-level contract, and then extract all that information by updating a variable at the "end" of the visitor when it gets to the final expression in the file? Then the last step of tying that information to the provided information can be a much flatter process that doesn't need to visit the whole tree.

For the

`bar :: (A -> B) = lam(a): b end`

case, I'd prefer to not change the meaning of `x :: A = e`, just on the principle of substitution.

In fact, for the same reason I think it's problematic that this program is an error:

```
foo :: (Number -> Number)
foo = lam(a): a end
foo("not-a-num")
```

while this one isn't:

```
foo :: (Number -> Number)
sneak = lam(a): a end
foo = sneak
foo("not-a-num")
```

That makes me want to only tie this to `fun`s themselves

**Comment by schanzer** [comment]
Sorry for the delay (we're training this week). 

I don't have the headspace to process Joe's implementation comments, but I'm happy with the proposal to tie these to functions. Joe is right that we never worry about exports in BS.

**Comment by blerner** [comment]
I think you misunderstood me, or I didn't explain myself well. How often do you want students to be able to "copy your contract sheet to the top of your file"? Those contracts are these whole-program things that I'm worried about identifying. Shriram likes that ability... And I tend to agree, that seems like a useful thing to be able to write.  (The reason I mentioned provide is because I was intending to use the same whole-file name-resolution logic to figure what meaning these contracts should have.)

So maybe the right semantics for "where can contracts appear?" is, "right before the group in which they're defined, or at the top of the file if they're imported values". And then we'll temporarily give an error for such imported value contracts?

**Comment by blerner** [comment]
**Working examples!**

Assume `is-even` is defined properly.  Here's `is-odd`, followed by a bogus call -- note the highlighted error location.  Also note that the examples block sits between the contract and the function, and scoping still works properly:
![image](https://user-images.githubusercontent.com/918464/29251474-6f85eb38-8023-11e7-8d4a-111bc0c64f6e.png)
![image](https://user-images.githubusercontent.com/918464/29251477-8088c720-8023-11e7-879c-5044dd561b57.png)
A contract whose argument names don't match the names of the function (same error appears if the arities don't match).  The cyan highlight is a lazy-mouseover highlight:
![image](https://user-images.githubusercontent.com/918464/29251491-9d90e0d2-8023-11e7-98cd-e741fbedb480.png)
Supplying a non-arrow contract to a syntactically-obvious function.  The highlight for `correspond` is a lazy-mouseover highlight:
![image](https://user-images.githubusercontent.com/918464/29251542-67824c1e-8024-11e7-8c87-dae5c664c58a.png)
Supplying an arrow contract to a syntactically-obvious non-function (again, `correspond` is lazily highlighted):
![image](https://user-images.githubusercontent.com/918464/29251549-983dec6e-8024-11e7-9a19-179a86a0e308.png)
Including `image`, then giving a contract to `circle`: (the blue highlight applies to the relevant `include` statement):
![image](https://user-images.githubusercontent.com/918464/29251554-abe108dc-8024-11e7-9a77-d6f96be27d9d.png)
Defining a contract on an annotated function:
![image](https://user-images.githubusercontent.com/918464/29251580-32eef384-8025-11e7-87b9-e1ee3caf67ab.png)
![image](https://user-images.githubusercontent.com/918464/29251581-37b0b8a8-8025-11e7-831e-d48a8cb36d61.png)
Defining a contract twice:
![image](https://user-images.githubusercontent.com/918464/29251654-b3652da2-8026-11e7-9230-8626d542c060.png)
![image](https://user-images.githubusercontent.com/918464/29251655-b86102c2-8026-11e7-8b83-f769698734ba.png)
Unused contract:
![image](https://user-images.githubusercontent.com/918464/29251661-d5491d0c-8026-11e7-8f24-12d8e80a5c43.png)
Contracts out of order in a letrec group:
![image](https://user-images.githubusercontent.com/918464/29251700-b776ea88-8027-11e7-98ca-f668fe1fbb1a.png)

**Comment by schanzer** [comment]
Wow. These look _awesome_, Ben! A few questions:
1) This doesn't effect Bootstrap, but I could see a lot of teachers giving kids a file with pre-written contracts that they have to implement. I'd love to leave this use-case open, if possible. The current implementation treats this as an error. Is this explicitly by design, or a logical consequence of the design? 
2) For contract whose argument names don't match the names of the function, do we have the luxury of giving separate errors for arity v. varnames? I'd love to be more specific in the error, if we can.
3) In the example where you supplying an arrow contract to a syntactically-obvious non-function (again, correspond is lazily highlighted), it took me several readings to grok the error. Can I suggest "`is-even` is given a function contract, but is not defined as a function"?

**Comment by blerner** [comment]
1. By design: I think it's more useful to know when a contract hasn't applied to anything. I'd rather the teacher give a black file and have the assignment worksheet lost the contracts.

2. Yes, we have the names available but I didn't include them in this draft. We probably need some pretty fancy matching to detect typos and (very likely) missing arguments, so I wasn't sure what UI I wanted there. That's @jswrenn's problem to solve 😎

3. Yeah, that could be a separate message.

**Comment by schanzer** [comment]
1. Last push I'll make, since this isn't a huge deal -- does CPO have a proper notion of _warnings_ v. _errors_? I'm assuming so, since test failures are sort of like warnings. If not, no big deal, but speaking as a teacher and curriculum designer, this would be really nice to have.

2. Cool!

3. Cool! :)

**Comment by jpolitz** [comment]
CPO doesn't. I'm usually pretty opposed to warnings, because they cause situations where we encourage people to ignore feedback to make progress.

This is one of the first really compelling cases for warnings I've seen, though – the "wrote a contract but didn't implement it yet" case. Because clearly the program should run in order to get feedback from tests, etc, but also the environment should clearly give feedback that the "TODO" list isn't done.

That's a useful use case to consider, though I agree that it's not a huge deal to worry about it for this issue.

**Comment by jpolitz** [comment]
Idea (again, not necessary for this impl) – what if a contract without an associated impl desugared to add a definition of the right arity and annotations, but with the body "..."

Then you could write the contract and the examples, and it would be well-formed to run the program and get a bunch of "NYI" errors in tests, without having to write out the `fun ... end` bit.

**Comment by blerner** [comment]
I like that idea, and considered it, but the hazard is, if a student screws up and has contracts at the top of the file that don't get woven onto anything, and later in the file they do have a definition, then my current implementation will complain that the contract needs to be adjacent to its associated definition, while this ... suggestion will lead to a shadowing error that IMO is frustratingly obtuse. To avoid that error, we'd need even more bookkeeping and AST enhancements that I don't think we want to do.

**Comment by jpolitz** [comment]
Could we do the adjacency check in well-formedness, before doing scope resolution? It seems _almost_ simple enough to put in well-formedness, though I understand that it interacts with let chains and examples, so it's not quite trivial.

**Comment by blerner** [comment]
I don't follow.  If we do it in well-formedness, then it's not guaranteed to be sound (we'd have to duplicate the logic of computing binding-groups in order to get it exactly right).  As it is, I'm already tightly coupling this to desugar-scope and desugar-scope-block, so I'm not sure how to make it any earlier without tripping ourselves.

Plus, unless students use the named-argument form of contracts, we wouldn't know enough to write the body without gensym'ing names.  And what would we show in a stack-trace, if we encountered an NYI body of these magick'ed up functions?

This feels like a slightly-too-cute idea to me, and it only saves students a minor amount of typing.  I'm going to reject it for this implementation, and maybe we can come back to it later down the line if it seems to be really compelling.

**Comment by jpolitz** [comment]
Agreed with the technical argument above, but I want to make one comment I feel is important:

I _don't_ believe that it's a “minor” amount of extra typing. Getting through `fun name(args): ... end` with no missed parens, colons, etc, is pretty hard. So one thing that's exciting to me about doing the desugaring is that it gives us a way to get to useful feedback about examples (they match the contract!) without having to type that in and deal with a bunch of syntax errors, which takes the student out of the process they were working on.

Still, not necessary for the PR referenced here, but something I'm getting more and more excited about as I think about this.

**Comment by blerner** [comment]
This feels more to me like an IDE action than a desugaring step. I *don't* think I want contracts to magically behave like nyi functions: it means we blur the line between "what is a definition of a function" and "what is the signature of a function", which is something we've seen students and teachers be very hazy on.

**Comment by jpolitz** [comment]
Regardless of where or how it's supposed to be implemented, I still think it's important, and helps with a real pain point in students going through the DR.

I think it could _help_ with the problem you mention, because it would make a step before defining the function itself quite concrete and checkable.
